### DISCUSSION TITLE
Elastic horovod: adding nodes does not work in certain situations

### DISCUSSION THREAD

**Discussion by sallyyuan1022** [horovod/horovod#2833] [Q&A]
**Environment:**

Framework: PyTorch
Framework version: 1.6.0
Horovod version: 0.21.3
MPI version: 4.0.3
CUDA version: 10.1
NCCL version: 2.5.6
Python version: 3.6.8
OS and version:
GCC version:

**Checklist:**
Did you search issues to find if somebody asked this question before?
If your question is about hang, did you read this doc?
If your question is about docker, did you read this doc?
Did you check if you question is answered in the troubleshooting guide?

**Your question:**
We are using elastic horovd with pytorch, in our usage scenarios, we often encounter such situations.

The _disvovery_hosts function detects that the host info has changed, but the coordinator info or worker client cannot be obtained in _notify_workers_host_changes func at the same time, resulting in the node update not working. (Both of the functions are in https://github.com/horovod/horovod/blob/master/horovod/runner/elastic/driver.py)

If the nodes do not change any more, these added nodes will never be used as training nodes.

We have tried to increase value of variable DISCOVER_HOSTS_FREQUENCY_SECS to decrease the frequency of calling _disvovery_hosts, but it is not a good solution.

Are there any good suggestions for dealing with this? 

Thanks,
Sally

**Comment by tgaddair** [comment]
Hey @sallyyuan1022, thanks for reporting this. This in fact appears to be a bug.  In the event that the worker client is unreachable, we should not update the state of the discovery process.  I'll create a PR for this.

**Comment by tgaddair** [comment]
Can you try #2838 and let me know if it addresses the issue?

**Reply by sallyyuan1022** [reply]
Hi @tgaddair, Sorry for replying you so late!

I feel that these changes would cause some other problems.

In the startup phase, before the worker has completed registration, the "_notify_workers_host_changes" method is sure to return false, so that "self._host_manager.current_hosts" will always be empty.
This should cause horovodrun to exit due to the timeout of waiting for available slots or min hosts in method "launch_gloo_elastic".

So I think the driver may need an attribute to indicate that horovodrun has workers in startup, which means driver.start() has been called. 

Hope for your suggestions!
